Refactor StrokeAlign to allow double values.

bernaferrari opened the pull request:
Issue: https://github.com/flutter/flutter/issues/106362

This is radical, this is breaking, this is awesome. cc @gspencergoog

Edit: technically it is not breaking, I only break a single test (that is dumb and might or might not be fixed by another PR). But StrokeAlign change is really big.

https://user-images.githubusercontent.com/351125/180920100-99b9e7cb-ee03-4f6c-8a9e-425c9c06b20c.mov

Also, I made a completely new sample:
<details>
  <summary>Click to see the sample code</summary>
  
```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData.dark(),
      home: const Scaffold(body: HomePage()),
    );
  }
}

class HomePage extends StatefulWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  State<HomePage> createState() => _HomePageState();
}

enum Shapes {
  roundedRectangleBorder,
  roundedinside20,
  roundedcenter30,
  roundedOutside40,
  circleBorder,
  ovalBorder,
  stadiumBorder
}

class _HomePageState extends State<HomePage> {
  double sliderValue = 0;
  Shapes from = Shapes.roundedinside20;
  Shapes to = Shapes.roundedOutside40;

  OutlinedBorder getBorder(Shapes shape) {
    switch (shape) {
      case Shapes.roundedRectangleBorder:
        return RoundedRectangleBorder(
          borderRadius: BorderRadius.circular(20),
          side: BorderSide(
              color: Colors.green.withOpacity(0.5),
              width: 20,
              strokeAlign: StrokeAlign.center),
        );
      case Shapes.roundedinside20:
        return RoundedRectangleBorder(
          borderRadius: BorderRadius.circular(5),
          side: BorderSide(
              color: Colors.purple.withOpacity(0.5),
              width: 20,
              strokeAlign: StrokeAlign.inside),
        );
      case Shapes.roundedcenter30:
        return RoundedRectangleBorder(
          borderRadius: BorderRadius.circular(10),
          side: BorderSide(
              color: Colors.blue.withOpacity(0.5),
              width: 30,
              strokeAlign: StrokeAlign.center),
        );
      case Shapes.roundedOutside40:
        return RoundedRectangleBorder(
          borderRadius: BorderRadius.circular(15),
          side: BorderSide(
              color: Colors.red.withOpacity(0.5),
              width: 40,
              strokeAlign: StrokeAlign.outside),
        );
      case Shapes.circleBorder:
        return CircleBorder(
          side: BorderSide(
            color: Colors.green.withOpacity(0.5),
            width: 20,
            strokeAlign: StrokeAlign.center,
          ),
        );
      case Shapes.ovalBorder:
        return OvalBorder(
          side: BorderSide(
            color: Colors.yellow.withOpacity(0.5),
            width: 20,
            strokeAlign: StrokeAlign.center,
          ),
        );
      case Shapes.stadiumBorder:
        return StadiumBorder(
          side: BorderSide(
              color: Colors.yellow.withOpacity(0.5),
              width: 20,
              strokeAlign: StrokeAlign.center),
        );
    }
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.center,
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        Container(
          width: 400,
          height: 200,
          decoration: ShapeDecoration(
            shape: getBorder(from).lerpTo(getBorder(to), sliderValue) ??
                RoundedRectangleBorder(),
            color: Colors.white,
          ),
        ),
        const SizedBox(height: 40),
        Slider(
            // divisions: 10,
            value: sliderValue,
            onChanged: (value) {
              setState(() {
                sliderValue = value;
              });
            }),
        Row(
          crossAxisAlignment: CrossAxisAlignment.center,
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Text("From"),
            SizedBox(width: 10),
            DropdownButton<Shapes>(
                value: from,
                items: [
                  for (final shape in Shapes.values)
                    DropdownMenuItem(
                      value: shape,
                      child: Text(shape.toString()),
                    ),
                ],
                onChanged: (value) {
                  setState(() {
                    from = value!;
                  });
                }),
            SizedBox(width: 20),
            Text("To"),
            SizedBox(width: 10),
            DropdownButton<Shapes>(
                value: to,
                items: [
                  for (final shape in Shapes.values)
                    DropdownMenuItem(
                      value: shape,
                      child: Text(shape.toString()),
                    ),
                ],
                onChanged: (value) {
                  setState(() {
                    to = value!;
                  });
                }),
            SizedBox(width: 10),
            Text("${(sliderValue * 100).toStringAsFixed(0)}%"),
          ],
        ),
      ],
    );
  }
}
```
</details>

gspencergoog commented:
This is great.  It's not technically breaking if it doesn't break any tests in customer_testing or in Google internal tests. Since it's a new addition, it probably won't.

gspencergoog commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -30,16 +30,91 @@ enum BorderStyle {
+  /// This constructor allows any value between 0.0 and 1.0.
+  const StrokeAlign(this.value)
+      : assert(value != null),
+        assert(value >= 0.0, 'StrokeAlign must be greater than or equal to 0.0.'),
```
```suggestion
        assert(value >= 0.0, 'StrokeAlign must be greater than or equal to 0.0. $value was supplied instead.'),
```

It's always good to put the value that failed in the assert.

gspencergoog commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -30,16 +30,91 @@ enum BorderStyle {
+
   /// The border is drawn on the inside of the border path.
-  inside,
+  static const StrokeAlign inside = StrokeAlign(0.0);
```
For me, it would be more intuitive to have negative one be "inside" and one be "outside". Then it's easy to know where it is relative to "center". But this works too, I just have to think of it as the proportion of the line that lies outside of the shape.

bernaferrari replied:
Both are fine. I did this so it worked with lerp.

Can you help here? How would you do with -1, 0, 1?
<img width="761" alt="image" src="https://user-images.githubusercontent.com/351125/181121455-a42a0b75-2ec3-4e71-8d9e-a23502e08329.png">

bernaferrari replied:
Here

gspencergoog replied:
If the value goes from -1 to 1, then it's this:

```dart
// Inside == -1 => side.width
// Center == 0 => side.width * 0.5
// Outside == 1 => 0
EdgeInsets.all(side.width * (1 - (1 + side.strokeAlign.value) / 2));
```

bernaferrari replied:
How do you solve this?

```dart
          final RRect borderRect = borderRadius.resolve(textDirection).toRRect(rect);
          final RRect inner = borderRect.deflate(ui.lerpDouble(side.width, 0, side.strokeAlign.value)!);
          final RRect outer = borderRect.inflate(ui.lerpDouble(0, side.width, side.strokeAlign.value)!);
          canvas.drawDRRect(outer, inner, paint);
```
From RoundedRectangleBorder. Needs to `deflate(side.width)` (inside), `deflate(side.width/2) AND inflate(side.width/2)` (center), `inflate(side.width)` outside.

gspencergoog replied:
```dart
final RRect inner = borderRect.deflate(side.width * (1 - side.strokeAlign.value) / 2);
final RRect outer = borderRect.inflate(side.width * (side.strokeAlign.value + 1) / 2);
```

bernaferrari replied:
The new formula is weird. It shrinks and grows

https://user-images.githubusercontent.com/351125/181131300-c43bcdf7-d38a-4402-b5b6-7ea8e0725cae.mov

bernaferrari replied:
Ok, I got it working. I feel legibility went to the drain, but I'm trusting it works:
<img width="573" alt="image" src="https://user-images.githubusercontent.com/351125/181132237-3f0f4e43-b501-4707-8a9f-20ea6c2d1913.png">

bernaferrari replied:
is there no lerp function that does that?

bernaferrari replied:
I'm making my own.

bernaferrari replied:
95% of the cases are the same lerps:
- lerp from zero to side.width
- lerp from side.width to zero

Now I understand what it is doing. Thanks.

gspencergoog commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -30,16 +30,91 @@ enum BorderStyle {
+    } else if (this == outside) {
+      return 'outside';
+    } else {
+      return '$value';
```
You could have this say "outside by $value" or something.

gspencergoog commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -30,16 +30,91 @@ enum BorderStyle {
+  int get hashCode => value.hashCode;
+
+  /// Lerp a side contained in a circle.
+  static double radiusLerp(Rect rect, BorderSide side) {
+    if (side.strokeAlign.value < 0.5) { // inside
+      return ui.lerpDouble((rect.shortestSide - side.width) / 2.0, rect.shortestSide / 2.0, 2 * side.strokeAlign.value)!;
+    } else if (side.strokeAlign.value == 0.5) {
+      return rect.shortestSide / 2.0; // center
+    } else { // outside
+      return ui.lerpDouble(rect.shortestSide / 2.0, (rect.shortestSide + side.width) / 2.0, 2 * side.strokeAlign.value - 0.5)!;
+    }
+  }
```
If you had the value go from -1 to 1, you could just have this like so, and not need the lerp at all:
```suggestion
  static double radiusLerp(Rect rect, BorderSide side) {
    return rect.shortestSide / 2 + side.strokeAlign.value * side.width;
  }
```

gspencergoog commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -30,16 +30,91 @@ enum BorderStyle {
+  }
+
+  /// Lerp a side contained in a [RRect].
+  static RRect rrectLerp(RRect rrect, BorderSide side) {
+    if (side.strokeAlign.value < 0.5) { // inside
+      return rrect.deflate(ui.lerpDouble(side.width / 2, 0, 2 * side.strokeAlign.value)!);
+    } else if (side.strokeAlign.value == 0.5) { // center
+      return rrect;
+    } else { // outside
+      return rrect.inflate(ui.lerpDouble(0, side.width / 2, 2 * side.strokeAlign.value - 0.5)!);
+    }
+  }
```
Similarly, this can be:

```suggestion
  static RRect rrectLerp(RRect rrect, BorderSide side) {
    return rrect.inflate(side.width * side.strokeAlign.value);
  }
```

(the inflate arg can be negative).

bernaferrari replied:
wowwww I never thought about this 😳

bernaferrari commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -30,16 +30,91 @@ enum BorderStyle {
 /// - [outside] provides zero padding, as stroke is drawn entirely outside.
-enum StrokeAlign {
+@immutable
+class StrokeAlign {
```
I don't know, I only reused what existed already. Ideally, it would be StrokeAlign.inside/center/outside/custom, but enum doesn't support custom.

Like, how would you initialize it in your current proposal?  Something like this?

<img width="687" alt="image" src="https://user-images.githubusercontent.com/351125/181121095-33097888-f824-4237-95b2-c5b9c17c750b.png">

gspencergoog replied:
Yep, exactly (although I'd still advocate -1 to 1 instead).

bernaferrari replied:
Just answer the other comment, how to lerp the borders, my brain is too melted for that lol

If you can get that, I'll get the -1, 0, 1.

bernaferrari replied:
I hate the new names. I think we can stay with the double, but I propose something in the middle (which can be confusing, open to discussion): 
<img width="816" alt="image" src="https://user-images.githubusercontent.com/351125/181125004-8c710fa8-f45f-4c15-ba44-0a48b6ba4d9e.png">

`StrokeAlign.inside` which is just a shortcut.
Because `BorderSide.strokeAlignInside` seems too hidden for me.

bernaferrari replied:
Update: what if StrokeAlign is closed and you can only call StrokeAlign.inside/center/outside which would work exactly like an enum except it gives you a double?

bernaferrari commented:
I just made some (many) changes based on your comments. I love the double strokeAlign, but I'm still not sure I want to get rid of StrokeAlign. Can't we have a global `StrokeAlign.inside = 0.0`? I think `BorderSide.strokeAlignInside` is too weird.

gspencergoog commented:
I don't think `BorderSide.strokeAlignInside` is that weird, we have lots of places where there are const defaults for double values, and it's a descriptive name.

And it seems like waaay overkill to have an entire class that effectively is a namespace for three doubles. It makes it more complicated to assign values to the `strokeAlign` value too:

```dart
BorderSide(strokeAlign: StrokeAlign(value));
```
vs
```dart
BorderSide(strokeAlign: value);
```

Seems more complicated for no benefit.

bernaferrari commented:
I simplified further, now: `BorderSide(strokeAlign: StrokeAlign.inside)` or `BorderSide(strokeAlign: 0.5)`. The opposite, strokeAlign: `BorderSide.strokeAlignInside` feels like a secret API that won't appear when typing strokeAlign, only borderSide.

gspencergoog commented:
One other thing to consider is that you might want to allow values outside of -1 and 1, so that you can animate the border with a bounce animation or other animation that goes slightly outside the bounds.  The equations I gave you should mostly just work for that, although if people go too crazy, it'll result in some odd behavior (rects that get deflated to zero size, for instance).

bernaferrari commented:
The math for -1, 0, 1 is very hard to understand, but I think I got it working. Need to test if getInnerPath is working correctly later.

bernaferrari commented:
I thought about that. I think we should set loose the constraints and wait to see what shows up. Technically there isn't a difference from changing that to changing the `width`, it is just augmenting it.

But, if people want to have fun... commiting.

gspencergoog commented:
Hey, @bernaferrari this looked like so much fun, I started playing with your code, and made my own branch with some changes. There's also some example code you might want to grab.

Here's my branch: https://github.com/gspencergoog/flutter/tree/stroke_align (and [the diff](https://github.com/gspencergoog/flutter/compare/master...gspencergoog:flutter:stroke_align?expand=1) if you'd rather look at that).

The sample looks like this when it runs:

https://user-images.githubusercontent.com/8867023/181156167-5f29d7d1-6fbe-48e2-9264-33a89467a194.mp4
https://user-images.githubusercontent.com/8867023/181156174-09199cf4-4cc8-4ada-8795-e2fe035f92e6.mp4

Not intending it as a replacement for yours, just some ideas for you, feel free to appropriate any/all/none of it.  Particularly useful, I think, were the new getters I added on `BorderSide` for `strokeInset`, `strokeOutset`, and `strokeOffset`.

bernaferrari commented:
I'll check soon (need to sleep), that's awesome! I told you it is fun.. lol this is most fun I've had in a while.

I still think you should rewrite ContinousRectangleBorder into something useful. It is the worst border by far, and it doesn't support any StrokeAlign 👀. Your Polygon code + this PR = a border everybody will love.

bernaferrari commented:
If you do this in your test, you will see some small weirdness:

`static final Color borderColor = Colors.red.shade500.withOpacity(0.50);`

![Uploading image.png…]()

gspencergoog commented:
> If you do this in your test, you will see some small weirdness

Yes, perhaps it needs some more work.

bernaferrari committed:
Refactor StrokeAlign to allow any value.

Refactor all OutlinedBorders.

Make everything work.

Add tests.

Move dimensions to OutlinedBorder so it can be reused.

PR refactoring.

Make StrokeAlign closed?

Add -1, 0, 1.

Unconstraint strokeAlign.

Adjust so everything works.

Fix analyzer.

bernaferrari committed:
Greg Patch

bernaferrari committed:
Patch fixes.

bernaferrari commented:
That bug also happenedd in my branch, forget it. I think it is a bug in `drawPath` rounding almost-square to square (or something like this). I've seen it happening in other places 😢.

Ok, so fantastic work!

`strokeInset` and `strokeOutset` are exactly what I wanted when I made the similar but bad named `lerpStrokeAlignDecreasing` lol.

I'm merging all of your branch here (* except for something I'm still too connected to let it go, today 🤣).

bernaferrari commented:
The `Diagnosticable` seems reasonably big. Do we expect a random number after BorderSide now?
<img width="735" alt="image" src="https://user-images.githubusercontent.com/351125/181351394-9cbfe24c-04ef-4448-b957-7239ce307567.png">

gspencergoog commented:
> The Diagnosticable seems reasonably big. Do we expect a random number after BorderSide now?

We use `Diagnosticable` on all the widgets, and many other classes, to handle formatting of the debug strings in a consistent way.  The "random number" is part of the object hash, so that you can tell if two instances of the object are the same instance or not.  It's not big in the runtime sense: it doesn't have any data members, and only costs something when `toString` is called, which is only used in debug builds anyhow.

bernaferrari commented:
Ok, I'll fix our tests, but I feel you will need to patch Google. That will probably be our breaking change.

bernaferrari commented:
Plot twist: hash is really random. Apparently enum.hash always doesn't perform ==/hashcode (like you would expect). Fixed using `style.name` (is this ok?).

bernaferrari commented:
I don't know how to hash things, shouldn't the hash value be always the same?
Due to this, Diagnosticable result is random.
If I do `Object.hash(1, 1)` in a test and execute from terminal twice, the results are different.

gspencergoog commented:
The Object hash includes a component for the base object instance's location in memory, and since the object has a new instance each time, it appears random.

bernaferrari commented:
Is it possible to test that? Because there some tests that expect `toString()`.

gspencergoog commented:
Yes, use `expect(foo.toString(), equalsIgnoringHashCodes('Expected#00000()'));`

bernaferrari commented:
How would that work here?
![image](https://user-images.githubusercontent.com/351125/181381226-d6c36986-74c1-4c52-9d0c-7d02aa3434a1.png)

gspencergoog commented:
Ahh.  It won't.  OK, that's annoying. Instead, just add this to `BorderSide`:

```dart
  @override
  String toStringShort() => 'BorderSide';
```

bernaferrari commented:
Thanks. Aaaaalmost there! 

So, I changed these to be the same default as BorderSide:
<img width="814" alt="image" src="https://user-images.githubusercontent.com/351125/181414008-b3b8a082-74cf-4634-ad2c-3c5340c12549.png">
But now it produces `BorderSide` (instead of `BorderSide()`) and `BorderSide(width: 0.0, style: none)` (instead of `style: BorderStyle.none`. Any ideas on how to solve it?

gspencergoog commented:
Those are all expected, and preferred, they're just different from the home-grown `toString` (which is why we should switch to using the `Diagnosticable` anyhow, since then they're all more consistent).

You might check and make sure that the `defaultValue` arguments in `debugFillProperties` have the correct defaults. Then it will automatically eliminate any parameters that are set to their default values, for brevity.

bernaferrari commented:
Ok! Nice. I'll update the tests. 

BTW, you forgot BoxDecoration uses `border/2`, ShapeDecoration uses `border`.  Fixed. hehe

gspencergoog commented:
Ahh, good catch.

bernaferrari commented:
I'm happy and sad this PR is almost ending. What will we do next?

bernaferrari commented:
I can be wrong, but I think `getOuterPath` shouldn't inflate/deflate, it is always the same (it is usually the painted area, so it shouldn't consider strokeAlign). Fixing.

bernaferrari committed:
Make dimensions non-negative.

Fix most tests.

bernaferrari committed:
Revert "Add multiple StrokeAlign."

This reverts commit 123468cba7f7676bafb3d6675ef106d8806ae0e5.

bernaferrari committed:
Add multiple StrokeAlign.

bernaferrari commented:
Ok, I just pushed the multi strokeAlign change. I made `strokeInset` a function, for now, so I can get `strokeInset / 2` in an easier way, as there is no `Rect / 2` (and dividing once is better than diving four times. Can I overload an operator as an extension? I don't know). Doing this would also allow us to see where that pattern is used and optimize with `strokeHalfInset` or something like that. 

Everything else is mostly the same. Most tests are passing. There are a few small unfinished touches here and there.

Do you think I should revert and land in a separate PR? Most of the changes are relatively small and involve replacing `inflate` with `inflateWithRect`.

gspencergoog commented:
> Do you think I should revert and land in a separate PR? 

Yes, I think that would be better: we're not sure yet that we want that functionality, and I wouldn't want that to delay the base alignment functionality, which seems generally useful.

bernaferrari commented:
Ok. I think there only two things left:

- What happens with RoundedRectangle + Border zero. Do we want it to 'jump' when lerping? Or just not be squared when radius = zero. With the multi StrokeAlign, I think it was better to keep rrect all the time.
- Deciding the `StrokeAlign` vs `BorderSide.strokeAlignInside` situation.
- StarBorder, I need to pull tomorrow.

bernaferrari commented:
Oh, BTW, do you want strokeAlign in ContinuousRectangleBorder? It has issues like this. Only middle (which is true by default) works well:

<img width="543" alt="image" src="https://user-images.githubusercontent.com/351125/181665952-7e4ea991-78da-445d-9764-1526b1920e87.png">

gspencergoog commented:
> Oh, BTW, do you want strokeAlign in ContinuousRectangleBorder?

It's OK to remove it for now. If I "fix" it, maybe it'll come back.

> Ok. I think there only two things left:

I think you have an "off-by-one" error, I count three things. :-)

> What happens with RoundedRectangle + Border zero. 

I'd say it should not jump in the lerp. And you're talking about the case where the strokeAlign is "outside", not the default case of "inside", right? Because I think that for rounded rects, it's expected that the rounded edge appears as it goes outside, since it inflates the rect. The problem is that if you do want an actual rect with an outside border, we don't have a border for that.

We could do some kind of crazy thing where we draw values less than 2 pixels differently from the rest of the values, making the "outside" alignments draw a progressively sharper rectangle as they approach zero. Then the lerp would really quickly go from square to round from 0 to 2.  That might be the best option, if we can work it out. It would need to match up neatly when it hits two pixels.

> Deciding the StrokeAlign vs BorderSide.strokeAlignInside situation. 

You know what I think about this. I say just nuke it and use a double. It can always come back later if needed.

> StarBorder

I can take care of that later if you want.

bernaferrari committed:
Small fixes.

Remove rect (always use rrect).
Fix smoke test.
Don't strokeAlign continuous_rectangle_border.

bernaferrari committed:
Fix StarBorder.

bernaferrari commented:
I'll let you win this fight, but the war isn't over. I hope to get my StrokeAlign back in the 2D StrokeAlign (I won't rest until we can to make it useful, lol).

StarBorder is done. Easy. I love this PR.

> The problem is that if you do want an actual rect with an outside border, we don't have a border for that

It could be BoxDecoration. What do you think? It already makes rect for `borderRadius = null`, could do for `borderRadius == 0`. Lerping is recommended for RoundedRectangleBorder anyway.

bernaferrari committed:
Remove StrokeAlign. Most painful commit in a while.

bernaferrari committed:
Fix typo.

bernaferrari commented:
Update sample with better color, star, oval, and removed ContinousRectangleBorder (until something happens 👀).

![image](https://user-images.githubusercontent.com/351125/181824559-2d89fc56-107f-4084-a935-133ce57601da.png)

bernaferrari committed:
Refactor sample.

bernaferrari committed:
Rename strokeOffset -> halfStrokeOffset

bernaferrari committed:
Imprrove documentation.

bernaferrari committed:
Fix doc.

gspencergoog commented:
> It could be BoxDecoration. What do you think? It already makes rect for borderRadius = null, could do for borderRadius == 0. Lerping is recommended for RoundedRectangleBorder anyway. Or we leave it there doing its thing when borderRadius = null.

Oh, yeah, BoxDecoration it is.  OK, then I think `RoundedRectangleBorder` should have a smooth lerp for sure then, and `RoundedRectangleBorder`s with outside stroke alignment will have curved corners, even if they have a zero borderRadius, because that's just the most consistent.

bernaferrari commented:
I think that's it, then. Feel free to review when you can. And prepare the g3fix, I think we will need it.

Final possible optimization: add `if(side.width == 0) return;` to every paint (before it was paint rect in RoundedRectangleBorder, you removed it). But BorderStyle.none already catches some cases, so I don't think that would make any difference. Not doing this change, I'm letting it paint an empty drrect.

bernaferrari committed:
Fix BoxDecoration with borderRadius.

bernaferrari commented:
[News form weekend] I'm suffering a bit without StrokeAlign. Autocomplete is not always trustworthy and prototyping is harder.

https://user-images.githubusercontent.com/351125/182002451-593e69ce-7567-4eef-96f1-2c47a77116c9.mov

bernaferrari committed:
Re-add StrokeAlign class. It is back. For now.

Remove `width == 0.0` so it follows the same pattern as others.

gspencergoog commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -151,20 +159,29 @@ class BorderSide {
-  /// The direction of where the border will be drawn relative to the container.
-  final StrokeAlign strokeAlign;
+  /// The relative position of the stroke on a [BorderSide] in a [Border] or [OutlinedBorder].
+  /// When set to [StrokeAlign.inside], the stroke is drawn completely inside the widget.
```
Needs a blank doc comment line after the first sentence (dartdoc uses that to make a summary line).

gspencergoog commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -151,20 +159,29 @@ class BorderSide {
+  /// [Container.clipBehavior] can be used in an outside widget to clip it.
+  /// If [Container.decoration] has a border, the container may incorporate
+  /// [width] as additional padding:
+  /// - [StrokeAlign.inside] provides padding with full [width].
```
I think first time readers of this will be confused by the reference to `StrokeAlign`, since it's not referenced in any other way. Their first thought on seeing that would probably be "Oh? What does the `StrokeAlign` class do, maybe I need that?"

I still prefer the regular doubles, and referring to `[strokeAlignInside]` instead, since it's a value on this same class.

bernaferrari replied:
I agree from the logical point on the `strokeAlignInside`. But development experience got so much worse for me. Typing `c`+enter is better when it doesn't add `canLerp`. shift + alt + back I only had to rewrite what I needed, now need a lot more.

I think this comment appeared when we decided to use doubles. It just occurred to me it is a copy of what is inside StrokeAlign. I'll try to reword a bit, let's see if we can make it work. If not, you know, we can go with strokeAlignInside, but I really would rather not.

I would be willing to

bernaferrari replied:
Changed to:
<img width="655" alt="image" src="https://user-images.githubusercontent.com/351125/182205327-b9ef6890-3d1b-49f8-8735-657da05d7ac4.png">

gspencergoog commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -294,6 +307,62 @@ class BorderSide {
+  /// This will return 0 for a [strokeAlign] of 1, and the width of the stroke for
+  /// a [strokeAlign] of -1.
+  ///
+  /// Simplified example:
```
What does this example show?  It doesn't seem related to `strokeInset`, and I'm not sure when I would copy and use that code as a developer.

bernaferrari replied:
I change it to:
```
/// Example of results:
/// StrokeAlign.inside (-1) => width
/// StrokeAlign.center (0) => width / 2
/// StrokeAlign.outside (1) => 0
```
What do you think? The main idea was to show how progressing the values change the output in a less-secret way.

gspencergoog commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -294,6 +307,62 @@ class BorderSide {
+  /// This will return 0 for a [strokeAlign] of -1, and the width of the stroke for
+  /// a [strokeAlign] of 1.
+  ///
+  /// Simplified example:
```
Same here, I don't understand the relevance.

gspencergoog commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -294,6 +307,62 @@ class BorderSide {
+  /// This will return half width of the stroke for a [strokeAlign] of 1, and the
+  /// negative half width for a [strokeAlign] of -1.
+  ///
+  /// Simplified example:
```
Same here.

bernaferrari commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -294,6 +323,30 @@ class BorderSide {
   }
 
+    /// Lerp a side contained in a circle.
+  static double radiusLerp(Rect rect, BorderSide side) =>
```
I believe you reviewed an outdated version, there is no radiusLerp anymore.

gspencergoog replied:
Sorry, I accidentally submitted an old comment.  You can ignore it.

bernaferrari commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -314,7 +367,7 @@ class BorderSide {
   @override
   String toString() {
-    if (strokeAlign == StrokeAlign.inside) {
+    if (strokeAlign == strokeAlignInside) {
```
This seems to be an old/previous version, there is no BorderSide.toString() anymore (only the one Diagnosticable generates).

gspencergoog replied:
I think this was a draft comment I accidentally left around, and then submitted. Thanks GitHub. :-)

You can ignore it.

bernaferrari committed:
Fix PR comments.

bernaferrari committed:
Fix analyzer.

gspencergoog commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -151,20 +162,32 @@ class BorderSide {
+  ///
+  /// Values typically range from -1.0 (inside border, default) to
+  /// 1.0 (outside border), without any bound constraints (e.g., a
+  /// value of -2.0 is unexpected, but possible).
```
```suggestion
  /// value of -2.0 is not typical, but allowed).
```

gspencergoog commented on packages/flutter/lib/src/painting/borders.dart:
````diff
@@ -294,6 +313,41 @@ class BorderSide {
     );
   }
 
+  /// Get the amount of the stroke width that lies inside of the [BorderSide].
+  ///
+  /// This will return 0 for a [strokeAlign] of 1, and the width of the stroke for
+  /// a [strokeAlign] of -1.
+  ///
+  /// Example of results:
+  /// StrokeAlign.inside (-1) => width
+  /// StrokeAlign.center  (0) => width / 2
+  /// StrokeAlign.outside (1) => 0
+  /// ```
````
It's better than the code, but I think just using the words might work better:

```suggestion
  /// Get the amount of the stroke width that lies inside of the [BorderSide].
  ///
  /// For example, this will return 0 for a [strokeAlign] of 1, half the [width] for
  /// a [strokeAlign] of 0, and the [width] for a [strokeAlign] of -1.
```

bernaferrari replied:
Last try. What do you think?
<img width="684" alt="image" src="https://user-images.githubusercontent.com/351125/182261754-915ff62a-9c4d-401b-944e-79a1344a59cb.png">

bernaferrari replied:
One line per strokeAlign.

bernaferrari replied:
Ok, give up. I'm doing what you recommend, but -1 -> 1 (the old was 1 -> -1). I don't understand anything, it is too hard to read. But I don't know how to improve.

gspencergoog replied:
Those look good now, and -1, 0, 1 seems logical.

bernaferrari committed:
Rename description

gspencergoog commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -294,6 +313,25 @@ class BorderSide {
+
+  /// Get the amount of the stroke width that lies outside of the [BorderSide].
+  ///
+  /// This will return 0 for a [strokeAlign] of -1, half the [width] for a
```
You might add the "For example, " at the beginning of these last two, just so they know that those aren't the only three values it returns.

bernaferrari replied:
Done

bernaferrari committed:
Add "for example".

bernaferrari committed:
Fix typo.

HansMuller commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -151,20 +162,32 @@ class BorderSide {
+  ///
+  /// See also:
+  ///
+  ///  * [StrokeAlign], a class that acts as a shortcut for specifying
```
I realize that this is a source of controversy, but I think the need to document the values of StrokeAlign's constants here supports the idea that we'd be better off with comparable constants defined on BorderSide.  Defining a class just to enshrine 3 constants that could be given equally evocative names and sufficient API doc here, doesn't seem like a good trade. 

That said, this a great PR; adds a nice dimension of flexibility to BorderSide.

bernaferrari replied:
Hey, Hans! Thanks for the review! I'm feeling the pressure for BorderSide to return. lol

I'll explain my thinking, if you disagree, I'll revert to `BorderSide.strokeAlignInside`, ok?
My rationale was:

Pros for StrokeAlign.inside:
- Do not break existing API (either way wouldn't be breaking, StrokeAlign is barely in Flutter beta and no one uses yet).
- Development experience is a lot easier. There is no autocomplete confusion, the word to type is much shorter, ctrl+shift+back + 'center' just works, co-pilot gets less on the way. If you type `c` it suggests `center` instead of `canLerp`.
- Discoverability is better. There is `style: BorderStyle.solid` and `color: Color()`, so `strokeAlign: StrokeAlign.inside` is natural.

Hypothetically, thinking about the future:
- If eventually, we go from `double` to `Offset` (I have the PR mostly ready for two-dimension StrokeAlign, but it lacks a reason to exist, so don't expect it soon), the migration path is enum -> double -> offset, with the same `StrokeAlign.inside` working always.  `BorderSide.strokeAlignInside` takes us to enum -> BorderSide -> (deprecation?) -> Offset.
- If we get the multi-directional BorderSide (PR almost ready, need to convince the team first it is a good idea), there are two possible implementations: `BorderSide.only()` (I prefer) and `MultiBorderSide extends BorderSide`. If we go with the second, `MultiBorderSide.strokeAlignInside` would be possible, which is also odd.

These two are pure speculation, I am just trying to look at what the future holds depending on the decision we take here.

Pros for `BorderSide.strokeAlignCenter`:
- You avoid allocating a class that doesn't need to exist.
- Theoretically should be slightlyyyy 0.0001s faster due to this.
- Behaves like other Flutter APIs (which is why I'm trying to go 2D and justify StrokeAlign lol).

I think everyone agrees `BorderSide.strokeAlignCenter` is ugly but more efficient. I think the same. We've gone forth and back a few times in this PR. I think a single class allocation could be forgiven due to the benefits in development performance. Is it 0.0001s slower? Possibly. But I think the development advtange pays the cost.

That said, do you still prefer `BorderSide.strokeAlignInside`?

bernaferrari replied:
2x1, but speed is important, so it is a tie. I'll let you decide 🤣.

<img width="523" alt="image" src="https://user-images.githubusercontent.com/351125/182285641-dfa50def-981c-471e-bb73-7bbb348d2e36.png">

HansMuller replied:
Thank you for the comprehensive analysis and for the nicely illustrated advantages in terms for Flutter's core values. I don't think I've seen a perspective on a technical decision expressed quite like this before.

I think a 2D strokeAlign parameter would enable some interesting effects, and it's a creative way to extend the stroke alignment idea, but I don't think it's the ideal way to generalize the definition of stroke width. That seems like it needs to be a function of distance along the border, like a Bezier curve's _t_ parameter. I don't think that kind of generalization is a feature that would benefit most developers. At least not now.

I do still prefer BorderSide.strokeAlignInside et al. I appreciate your passion about this feature and I hope we can make this small adjustment and get this PR landed. It's a nice piece of work.

bernaferrari replied:
>  That seems like it needs to be a function of distance along the border, like a Bezier curve's t parameter.

That sounds fascinating, could you explain more? Maybe I can try later. Right now, the 2D StrokeAlign is a different StrokeAlign for each side. So you can have left side set to -1, right side to center (0), and right side to outside (1). It works well, symmetric shapes can make a nice effect. Maybe I can start a document proposal and we can think together. Another day. Right now I'm writing [this proposal](https://docs.google.com/document/d/1rEQhlPzUj1QPEGSiA7eLl6kEqlkJFUecpKn-9Hcsddc/edit?usp=sharing&resourcekey=0-fpMavpYbDkqrkW2iNclV_Q). Hixie didn't like `BorderSide.only`, so I'm thinking about rewriting Border (and if it is well accepted, then fix BorderSide or maybe never fix it). Feel free to help if you like the idea. Greg has had enough of me already.

> I do still prefer BorderSide.strokeAlignInside
Ok :(
You won. Reverted.

bernaferrari committed:
Bring back BorderSide.strokeAlignInside.

bernaferrari committed:
Fix test

bernaferrari commented:
This one is ready. Unfortunately, BorderSide is back. I also updated some comments regarding its return.

gspencergoog commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -151,20 +132,48 @@ class BorderSide {
+  /// The relative position of the stroke on a [BorderSide] in an
+  /// [OutlinedBorder] or [Border].
+  ///
+  /// Values typically range from -1.0 (inside border, default) to
+  /// 1.0 (outside border), without any bound constraints (e.g., a
+  /// value of -2.0 is is not typical, but allowed).
```
I'd add mentions of the constants here, so the reader knows what they correspond to.

```suggestion
  /// Values typically range from -1.0 ([strokeAlignInside], inside border, default) to
  /// 1.0 ([strokeAlignOutside] outside border), without any bound constraints (e.g., a
  /// value of -2.0 is is not typical, but allowed). A value of 0 ([strokeAlignCenter]) will
  /// center the border on the edge of the widget.
```

gspencergoog commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -294,6 +299,26 @@ class BorderSide {
+  /// of 1.
+  double get strokeOutset => width * (1 + strokeAlign) / 2;
+
+  /// The offset of the stroke, taking into account the stroke alignment.
```
Isn't it half the offset?

Incidentally, as a public API, it's not quite clear why this is a half offset, or why I'd need that.  I think it might be better to just call it `strokeOffset` and then divide it by two where you use it: that value is more interesting. Yes, it's more typing in your code, but I think it makes for a more useful, clear API.

I mean, really, there's not really a huge need for this accessor anyhow, it's just convenient for implementing borders.

gspencergoog commented on packages/flutter/lib/src/painting/continuous_rectangle_border.dart:
```diff
@@ -128,6 +128,7 @@ class ContinuousRectangleBorder extends OutlinedBorder {
 
   @override
   Path getOuterPath(Rect rect, { TextDirection? textDirection }) {
+    // Has to deflate first to get the right radius when we inflate again.
```
Is this comment still relevant?

gspencergoog commented on packages/flutter/test/painting/rounded_rectangle_border_test.dart:
```diff
@@ -97,28 +97,28 @@ void main() {
     expect(
       ShapeBorder.lerp(r, c, 0.1).toString(),
-      'RoundedRectangleBorder(BorderSide(Color(0xff000000), 0.0, BorderStyle.none), BorderRadius.circular(10.0), 10.0% of the way to being a CircleBorder)',
+      equalsIgnoringHashCodes('RoundedRectangleBorder(BorderSide(width: 0.0, style: none), BorderRadius.circular(10.0), 10.0% of the way to being a CircleBorder)'),
```
These don't need to use `equalsIgnoringHashCodes` anymore.

BTW, I did land a change that allows `equalsIgnoringHashCodes` to take a List<String> or a string.

gspencergoog commented on examples/api/lib/painting/borders/border_side.0.dart:
```diff
@@ -0,0 +1,204 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
This example needs a test.  See some of the other tests in examples/api/test for guidance. They're pretty simple tests: you could just verify that the expected `TestBox`es are in the widget tree after you pump a `BorderSideExample`.

gspencergoog reviewed (changes requested):
This is looking really good.  Just some small things and we can get it in!

bernaferrari committed:
Fix PR comments.

gspencergoog reviewed (approved):
![32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d](https://user-images.githubusercontent.com/8867023/37737237-66e9c9dc-2d10-11e8-8797-13a405eace49.png)

Yay!

bernaferrari commented:
I guess we ping @HansMuller now? Ping.

HansMuller reviewed (approved):
LGTM

bernaferrari committed:
Fix redundant import.

bernaferrari committed:
Fix test (ops).

CaseyHillers commented:
FYI this broke many tests, I've uploaded a revert in https://github.com/flutter/flutter/pull/109591. @gspencergoog and @HansMuller can you help with the reland to prevent these breakages?

bernaferrari commented:
How bad is it? Is it a golden, a bug from the code, the StrokeAlign or something else?

HansMuller commented:
This PR removed the StrokeAlign enum. That's a breaking change and it caused many internal test failures due to a StrokeAlign dependency (and see https://docs.flutter.dev/resources/compatibility). In this case a small G3Fix would probably enable this PR to land.

Failures reported in Google-internal issue b/242668688

gspencergoog commented:
OK, I'll take care of re-landing it.
